roscala/src/main/scala/coop/rchain/rosette/Ob.scala

```diff
@@ -10,38 +12,149 @@ case object Upcall extends LookupError
 
 trait Base
 
+//TODO change type of `indirect` argument to bool
 trait Ob extends Base {
-  val parent: Ob
   val meta: Ob
+  val parent: Ob
   val slot: Seq[Ob]
   val obTag: ObTag = null
   val sysval: SysCode = null
+  val constantP = true
 
   def dispatch(ctxt: Ctxt): Ob = null
   def extendWith(keymeta: Ob): Ob = null
   def extendWith(keymeta: Ob, argvec: Tuple): Ob = null
-  def getAddr(ind: Int, level: Int, offset: Int): Ob = null
+
+  def getAddr(ind: Int, level: Int, offset: Int): Ob = getLex(ind, level, offset) //TODO FIXNUM(ADDR_TO_PRE_FIXNUM _)
+
   def getField(ind: Int,
                level: Int,
                offset: Int,
                spanSize: Int,
-               sign: Int): Ob =
-    null
-  def getLex(ind: Int, level: Int, offset: Int): Ob = null
+               sign: Int): Ob = {
+
+    ??? //TODO
+  }
+
+  def getLex(ind: Int, level: Int, offset: Int): Ob = {
+    val p: Ob = nthParent(level)
+
+    actorExtensionFiltered(ind, p)
+      .map(offsetOrInvalid(offset, _))
+      .getOrElse(Ob.INVALID)
+  }
+
   def is(value: Ob.ObTag): Boolean = true
   def lookupOBO(meta: Ob, ob: Ob, key: Ob): Either[LookupError, Ob] =
     Right(null)
   def matches(ctxt: Ctxt): Boolean = false
   def numberOfSlots(): Int = Math.max(0, slot.length - 2)
   def runtimeError(msg: String, state: VMState): (RblError, VMState) =
     (DeadThread, state)
-  def setAddr(ind: Int, level: Int, offset: Int, value: Ob): Ob = null
+
+  def setAddr(ind: Int, level: Int, offset: Int, value: Ob): Ob = setLex(ind, level, offset, value) //TODO
+
   def setField(ind: Int,
                level: Int,
                offset: Int,
                spanSize: Int,
-               value: Int): Ob = null
-  def setLex(ind: Int, level: Int, offset: Int, value: Ob): Ob = null
+               value: Int): Ob = {
+    ??? //TODO
+  }
+
+  def setLex(ind: Int, level: Int, offset: Int, value: Ob): Ob = {
+    val p: Ob = nthParent(level)
+
+    actorExtensionFiltered(ind, p)
+      .filter { _ => offset < p.numberOfSlots() }
+      .map { ob =>
+        ob.slot(offset) = value
```

Comment:
We're using lenses to avoid mutation.  Talk to Timm about how to do this.

Reply:
```
import Lenses._

ob.update(_ >> 'slot)(_.updated(offset, value))
```
If the data structure is not deeply nested, I think `ob.copy(...)` is fine though.

The problem here however is that `ASSIGN(p, slot(offset), val);`/`ob.slot(offset) = value` is a side effects to `this`. Therefore `setLex` needs to return `(Ob, Ob)` or we allow for mutation in `Ob` for now until we figure things out.
I would argue for the latter.